### PR TITLE
New nzchar_linter

### PR DESCRIPTION
**Pull request by MichaelChirico**
Part of #884

Some notes:

 - We could easily make the behavior we chose to skip linting `if (x == "")` optional
 - This linter may be somewhat controversial because of the difference in handling `NA_character_`. I really find the mismatch of defaults in R confusing here. We could consider a weaker version of this linter that only lints `nchar(x, keepNA = FALSE)`
 - No lints on {lintr}

### CODE REVIEW

File: R/nzchar_linter.R [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+nzchar_linter <- function() {
+  # use string-length to capture both "" and ''
+  # if (any(x == "")) is not treated like it's part of if(), but
+  #   any(if (x == "") y else z) _is_ treated so. this condition looks for the
+  #   expr to be inside a call that's _not_ above an IF/WHILE.
+  comparison_xpath <- "
+  //STR_CONST[string-length(text()) = 2]
+    /parent::expr
+    /parent::expr[
+      (EQ or NE or LT or GT or LE or GE)
```
Comment (AshesITR) [inline R/nzchar_linter.R:23]: Why lint `<, >, <=, >=`?

File: R/nzchar_linter.R [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  "
+
+  # nchar(., type="width") not strictly compatible with nzchar
+  # unsure allowNA compatible, so allow it just in case (see TODO in tests)
+  nchar_xpath <- "
+  //SYMBOL_FUNCTION_CALL[text() = 'nchar']
+    /parent::expr
+    /parent::expr
+    /parent::expr[
+      (EQ or NE or LT or GT or LE or GE)
```
Comment (AshesITR) [inline R/nzchar_linter.R:48]: Of these, only `>`, `==`, `!=` and `<=` (which is `==` in disguise) make sense. `<` is always false, `>=` is always true.

File: R/nzchar_linter.R [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+    xml <- source_expression$xml_parsed_content
+
+    comparison_expr <- xml_find_all(xml, comparison_xpath)
+    comparison_lints <- xml_nodes_to_lints(
+      comparison_expr,
+      source_expression = source_expression,
+      lint_message = paste(
+        'Instead of comparing strings to "", use nzchar().',
+        'An exception is carved out for comparing scalar strings to "" in',
```
Comment (AshesITR) [inline R/nzchar_linter.R:81]: I wouldn't mention the exception in the lint message.

**Review by AshesITR — COMMENTED** [review]
Missing metadata tests.

File: R/nzchar_linter.R (hunk 2 above)
Comment (MichaelChirico) [reply R/nzchar_linter.R:48]: Right... feels fine to include those anyway. `x < ""` is a code smell I guess, seems OK to lint it here. Maybe with different lint metadata?

File: R/nzchar_linter.R (hunk 1 above)
Comment (MichaelChirico) [reply R/nzchar_linter.R:23]: Happens very rarely but does seem like it's done. Maybe as a little code golf thing?

https://github.com/search?q=lang%3AR+%2F%5B%5E%23%3D-%5D%3E%5Cs*%22%22%2F&type=code

Comment (AshesITR) [reply R/nzchar_linter.R:23]: Interesting, there are even some hits for `<=`. Thanks for checking.

File: R/nzchar_linter.R (hunk 2 above)
Comment (AshesITR) [reply R/nzchar_linter.R:48]: This one is `nchar(x) < 0` // `nchar(x) >= 0`

Comment (MichaelChirico) [reply R/nzchar_linter.R:48]: 😅 guessed wrong from the preview. Still feels harmless to include (and I can switch to use `infix_metadata`), WDYT?

File: R/expect_comparison_linter.R [hunk 4]
```diff
@@ -49,7 +49,7 @@
 #' @export
 expect_comparison_linter <- function() {
   # != doesn't have a clean replacement
-  comparator_nodes <- setdiff(as.list(infix_metadata$xml_tag[infix_metadata$comparator]), "NE")
```
Comment (AshesITR) [inline R/expect_comparison_linter.R:52]: What's with this diff?

File: tests/testthat/test-nzchar_linter.R [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+})
+
+test_that("multiple lints are generated correctly", {
+  expect_lint(
+    trim_some("{
+      a == ''
+      nchar(b) != 0
+    }"),
+    list(
+      rex::rex('Instead of comparing strings to ""'),
```
Comment (AshesITR) [inline tests/testthat/test-nzchar_linter.R:69]: Add line number expectations here?

File: R/expect_comparison_linter.R (hunk 4 above)
Comment (MichaelChirico) [reply R/expect_comparison_linter.R:52]: I think it's a vestige from before `xp_or()` supported vector input (585e7cd3e0e4a3aa8f78fdf0242a908ff4cfd65e0). And `setdiff(list(), "NE")` looks weird to me.

**Review by AshesITR — DISMISSED** [review]

**Comment by IndrajeetPatil** [comment]
@MichaelChirico This should be ready for a merge after the merge conflict is resolved.

**Review by IndrajeetPatil — APPROVED** [review]

**Pull request merged**